File: go/pkg/buffer/ring_buffer.go
package buffer

import (
	"fmt"
	"io"
	"slices"
	"sync"
)

// RingBuffer is a thread-safe ring buffer that implements io.Reader and io.Writer
// interfaces. Unlike other buffer types, RingBuffer overwrites the oldest data
// when the buffer is full, making it suitable for maintaining a sliding window
// of the most recent data.
//
// The buffer uses head and tail pointers to implement a circular buffer.
// When the buffer is full, new writes overwrite the oldest data and advance
// the head pointer. When the buffer is empty, read operations block until
// data is written.
type RingBuffer[T any] struct {
	writeNotify chan struct{}

	mu         sync.Mutex
	buf        []T
	head, tail int64
	closeWrite bool
	closeErr   error
}

// RingN creates a new RingBuffer with the specified size.
// The buffer will overwrite the oldest data when this capacity is exceeded.
func RingN[T any](size int) *RingBuffer[T] {
	return &RingBuffer[T]{
		writeNotify: make(chan struct{}, 1),

		buf: make([]T, size),
	}
}

// Discard removes and discards the next n elements from the buffer without reading them.
// If n is greater than the available elements, all available data is discarded.
func (rb *RingBuffer[T]) Discard(n int) (err error) {
	rb.mu.Lock()
	defer rb.mu.Unlock()
	if rb.closeErr != nil {
		return fmt.Errorf("buffer: skip from closed buffer: %w", rb.closeErr)
	}
	if n > int(rb.tail-rb.head) {
		rb.head = rb.tail
		return nil
	}
	rb.head += int64(n)
	return nil
}

// Read reads data from the buffer into the provided slice.
// It blocks until data is available or the buffer is closed.
// Returns the number of elements read and any error encountered.
func (rb *RingBuffer[T]) Read(p []T) (int, error) {
	rb.mu.Lock()
	defer rb.mu.Unlock()

	if rb.closeErr != nil {
		return 0, fmt.Errorf("buffer: read from closed buffer: %w", rb.closeErr)
	}

	for rb.head == rb.tail {
		if rb.closeWrite {
			return 0, io.EOF
		}
		rb.mu.Unlock()
		<-rb.writeNotify
		rb.mu.Lock()
		if rb.closeErr != nil {
			return 0, fmt.Errorf("buffer: read from closed buffer: %w", rb.closeErr)
		}
	}

	avail := int(rb.tail - rb.head)
	head := int(rb.head % int64(len(rb.buf)))

	var n int
	if head+avail <= len(rb.buf) {
		n = copy(p, rb.buf[head:head+avail])
	} else {
		n = copy(p, rb.buf[head:])
		n += copy(p[n:], rb.buf[:avail-n])
	}

	rb.head += int64(n)
	return n, nil
}

// CloseWithError closes the buffer with the specified error.
// All pending operations are unblocked and return this error.
func (rb *RingBuffer[T]) CloseWithError(err error) error {
	if err == nil {
		err = io.ErrClosedPipe
	}
	rb.mu.Lock()
	defer rb.mu.Unlock()
	return rb.closeWithErrorLocked(err)
}

func (rb *RingBuffer[T]) closeWithErrorLocked(err error) error {
	if rb.closeErr != nil {
		return nil
	}
	rb.closeErr = err

	if !rb.closeWrite {
		rb.closeWrite = true
		close(rb.writeNotify)
	}
	return nil
}

// Error returns the error that caused the buffer to be closed, if any.
func (rb *RingBuffer[T]) Error() error {
	rb.mu.Lock()
	defer rb.mu.Unlock()
	return rb.closeErr
}

// Close closes the buffer, preventing further writes and reads.
// Equivalent to CloseWithError(io.ErrClosedPipe).
func (rb *RingBuffer[T]) Close() error {
	return rb.CloseWithError(io.ErrClosedPipe)
}

// CloseWrite closes the write side of the buffer, preventing further writes.
// Reads will continue to work until the buffer is empty, then return EOF.
func (rb *RingBuffer[T]) CloseWrite() error {
	rb.mu.Lock()
	defer rb.mu.Unlock()
	if rb.closeWrite {
		return nil
	}
	rb.closeWrite = true
	close(rb.writeNotify)
	return nil
}

// Write writes data from the provided slice to the buffer.
// Unlike other buffer types, RingBuffer overwrites the oldest data when full,
// rather than blocking. This ensures that the most recent data is always preserved.
// Returns the number of elements written and any error encountered.
func (rb *RingBuffer[T]) Write(p []T) (int, error) {
	rb.mu.Lock()
	defer rb.mu.Unlock()
	if rb.closeErr != nil {
		return 0, fmt.Errorf("buffer: write to closed buffer: %w", rb.closeErr)
	}
	if rb.closeWrite {
		return 0, fmt.Errorf("buffer: write to closed buffer: %w", io.ErrClosedPipe)
	}

	// write to full
	bufsz := int64(len(rb.buf))
	avail := int(bufsz - (rb.tail - rb.head))
	tail := int(rb.tail % bufsz)

	var wn int
	if avail > 0 {
		if tail+avail <= len(rb.buf) {
			wn = copy(rb.buf[tail:tail+avail], p)
		} else {
			wn = copy(rb.buf[tail:], p)
			wn += copy(rb.buf[:avail-wn], p[wn:])
		}
		rb.tail += int64(wn)
	}

	leftn := len(p) - wn
	if leftn == 0 {
		return wn, nil
	}

	// Now the buffer is full, and all we write the left `leftn` bytes to the
	// buffer. We can calculate the outcome result as follows:
	//
	//  [buffer]
	//   ......
	//   ......
	//   ....bb
	//   cccc
	//
	// Where 'b' and 'c' are the data we actually write. So we just skip the
	// first `.` bytes, and then copy `b` and `c` to the buffer.

	var cbuf, bbuf []T

	if leftn <= len(rb.buf) {
		//  [buffer]
		//   cccc
		cbuf = p[len(p)-leftn:]
	} else {
		//  [buffer]
		//   ......
		//   ......
		//   ....bb
		//   cccc
		cn := leftn % len(rb.buf)                // the number of `c`
		cbuf = p[len(p)-cn:]                     // the data of `c`
		bbuf = p[len(p)-len(rb.buf) : len(p)-cn] // the data of `b`
	}

	// now we need to copy the data to the buffer, and then copy the data to the
	// buffer again.
	//
	// tail - head must be equal to buffer size, as we already write a full
	// buffer in the previous step.
	head := int(rb.head % bufsz)
	if cp1 := copy(rb.buf[head:], cbuf); cp1 < len(cbuf) {
		//  The original buffer: 	......
		//  The head and tail:      	h....t
		//  The result should be:	ccbbcc
		cp2 := copy(rb.buf, cbuf[cp1:])
		copy(rb.buf[cp2:], bbuf)
	} else {
		//  The original buffer: 	......
		//  The head and tail:       h....t
		//  The result should be:	bccccb
		bp1 := copy(rb.buf[head+cp1:], bbuf)
		copy(rb.buf, bbuf[bp1:])
	}

	// move the head to `b` position.
	rb.head += int64(len(cbuf))
	rb.tail += int64(len(cbuf))

	return len(p), nil
}

// Next reads and returns the next element from the buffer.
// It blocks until an element is available or the buffer is closed.
// Returns ErrIteratorDone when the buffer is closed and empty.
func (rb *RingBuffer[T]) Next() (t T, err error) {
	rb.mu.Lock()
	defer rb.mu.Unlock()
	if rb.closeErr != nil {
		err = fmt.Errorf("buffer: read from closed buffer: %w", rb.closeErr)
		return
	}
	for rb.head == rb.tail {
		if rb.closeWrite {
			err = ErrIteratorDone
			return
		}
		rb.mu.Unlock()
		<-rb.writeNotify
		rb.mu.Lock()
		if rb.closeErr != nil {
			err = fmt.Errorf("buffer: read from closed buffer: %w", rb.closeErr)
			return
		}
	}
	head := rb.head % int64(len(rb.buf))
	chunk := rb.buf[head]
	rb.head++
	return chunk, nil
}

// Add adds a single element directly to the buffer.
// If the buffer is full, it will overwrite the oldest element and advance
// the head pointer to maintain the ring buffer behavior.
func (rb *RingBuffer[T]) Add(t T) error {
	rb.mu.Lock()
	defer rb.mu.Unlock()
	if rb.closeErr != nil {
		return fmt.Errorf("buffer: write to closed buffer: %w", rb.closeErr)
	}
	if rb.closeWrite {
		return fmt.Errorf("buffer: write to closed buffer: %w", io.ErrClosedPipe)
	}
	tail := rb.tail % int64(len(rb.buf))
	rb.buf[tail] = t
	rb.tail++
	if rb.tail-rb.head > int64(len(rb.buf)) {
		rb.head++
	}
	select {
	case rb.writeNotify <- struct{}{}:
	default:
	}
	return nil
}

// Reset clears the buffer by resetting the head and tail pointers.
// This effectively discards all buffered data.
func (rb *RingBuffer[T]) Reset() {
	rb.mu.Lock()
	defer rb.mu.Unlock()
	rb.head = 0
	rb.tail = 0
}

// Len returns the number of elements currently in the buffer.
func (rb *RingBuffer[T]) Len() int {
	rb.mu.Lock()
	defer rb.mu.Unlock()
	return int(rb.tail - rb.head)
}

// Bytes returns a slice containing all elements currently in the buffer.
// The returned slice is a copy of the buffered data.
func (rb *RingBuffer[T]) Bytes() []T {
	rb.mu.Lock()
	defer rb.mu.Unlock()
	h := rb.head % int64(len(rb.buf))
	t := rb.tail % int64(len(rb.buf))
	if h < t {
		return rb.buf[h:t]
	}
	return slices.Concat(rb.buf[h:], rb.buf[:t])
}

File: rust/buffer/src/ring_buffer.rs
//! Overwriting ring buffer implementation.

use std::error::Error;
use std::sync::{Arc, Condvar, Mutex};

use crate::error::{BufferError, Done};

/// A thread-safe overwriting ring buffer.
///
/// `RingBuffer<T>` is a thread-safe circular buffer with a fixed capacity.
/// When the buffer is full, new writes overwrite the oldest data instead of
/// blocking. This makes it ideal for maintaining a sliding window of the
/// most recent data.
///
/// # Semantics
///
/// - **Read**: Blocks when empty, returns data when available
/// - **Write**: Never blocks, overwrites oldest data when full
/// - **Close**: `close_write()` allows draining, `close_with_error()` immediate
///
/// # Example
///
/// ```
/// use giztoy_buffer::RingBuffer;
///
/// // Keep only the 100 most recent samples
/// let buf = RingBuffer::<f32>::new(100);
///
/// // Write samples - old data automatically overwritten
/// for i in 0..200 {
///     buf.add(i as f32).unwrap();
/// }
///
/// // Buffer contains only the last 100 samples (100..200)
/// assert_eq!(buf.len(), 100);
/// ```
pub struct RingBuffer<T> {
    inner: Arc<RingBufferInner<T>>,
}

struct RingBufferInner<T> {
    state: Mutex<RingBufferState<T>>,
    write_notify: Condvar,
}

struct RingBufferState<T> {
    buf: Vec<Option<T>>,
    // Virtual counters that track total read/write positions. These grow
    // monotonically and will wrap around after 2^64 operations on 64-bit
    // systems (2^32 on 32-bit). The implementation uses wrapping arithmetic
    // to handle this correctly.
    head: usize, // read position (virtual counter)
    tail: usize, // write position (virtual counter)
    close_write: bool,
    close_err: Option<Arc<dyn Error + Send + Sync>>,
}

impl<T> Clone for RingBuffer<T> {
    fn clone(&self) -> Self {
        RingBuffer {
            inner: Arc::clone(&self.inner),
        }
    }
}

impl<T> RingBuffer<T> {
    /// Creates a new RingBuffer with the specified capacity.
    ///
    /// The buffer will overwrite the oldest data when this capacity is exceeded.
    pub fn new(capacity: usize) -> Self {
        assert!(capacity > 0, "capacity must be greater than 0");
        let mut buf = Vec::with_capacity(capacity);
        buf.resize_with(capacity, || None);

        RingBuffer {
            inner: Arc::new(RingBufferInner {
                state: Mutex::new(RingBufferState {
                    buf,
                    head: 0,
                    tail: 0,
                    close_write: false,
                    close_err: None,
                }),
                write_notify: Condvar::new(),
            }),
        }
    }

    /// Returns the number of elements currently in the buffer.
    pub fn len(&self) -> usize {
        let state = self.inner.state.lock().unwrap();
        // Use wrapping_sub to handle potential overflow correctly
        state.tail.wrapping_sub(state.head).min(state.buf.len())
    }

    /// Returns the buffer capacity.
    pub fn capacity(&self) -> usize {
        let state = self.inner.state.lock().unwrap();
        state.buf.len()
    }

    /// Returns true if the buffer is empty.
    pub fn is_empty(&self) -> bool {
        let state = self.inner.state.lock().unwrap();
        state.head == state.tail
    }

    /// Resets the buffer by clearing all data.
    ///
    /// This does not change the closed state of the buffer.
    pub fn reset(&self) {
        let mut state = self.inner.state.lock().unwrap();
        for slot in &mut state.buf {
            *slot = None;
        }
        state.head = 0;
        state.tail = 0;
    }

    /// Returns the error that caused the buffer to be closed, if any.
    pub fn error(&self) -> Option<Arc<dyn Error + Send + Sync>> {
        let state = self.inner.state.lock().unwrap();
        state.close_err.clone()
    }

    /// Closes the write side of the buffer.
    ///
    /// This prevents new writes but allows existing data to be read.
    pub fn close_write(&self) -> Result<(), BufferError> {
        let mut state = self.inner.state.lock().unwrap();
        if state.close_write {
            return Ok(());
        }
        state.close_write = true;
        self.inner.write_notify.notify_all();
        Ok(())
    }

    /// Closes the buffer with the specified error.
    ///
    /// All blocking operations are immediately unblocked and return the error.
    pub fn close_with_error<E>(&self, err: E) -> Result<(), BufferError>
    where
        E: Error + Send + Sync + 'static,
    {
        let mut state = self.inner.state.lock().unwrap();
        if state.close_err.is_some() {
            return Ok(());
        }
        state.close_err = Some(Arc::new(err));
        if !state.close_write {
            state.close_write = true;
        }
        self.inner.write_notify.notify_all();
        Ok(())
    }

    /// Closes the buffer.
    ///
    /// This is semantically equivalent to `close_write()`, allowing readers to
    /// drain any remaining items while preventing further writes.
    pub fn close(&self) -> Result<(), BufferError> {
        self.close_write()
    }
}

impl<T: Clone> RingBuffer<T> {
    /// Writes data to the buffer.
    ///
    /// If the buffer is full, overwrites the oldest data.
    /// Write operations never block (except when closed).
    /// Returns the number of elements written (always `data.len()` on success).
    pub fn write(&self, data: &[T]) -> Result<usize, BufferError> {
        let mut state = self.inner.state.lock().unwrap();

        // Check for errors
        if let Some(ref err) = state.close_err {
            return Err(BufferError::ClosedWithError(Arc::clone(err)));
        }
        if state.close_write {
            return Err(BufferError::Closed);
        }

        let capacity = state.buf.len();

        // Write each item to the buffer. When the buffer is full, we overwrite
        // the oldest item by advancing head. This maintains the invariant that
        // tail.wrapping_sub(head) <= capacity at all times.
        for item in data {
            let tail_idx = state.tail % capacity;
            state.buf[tail_idx] = Some(item.clone());
            state.tail = state.tail.wrapping_add(1);

            // If buffer is overfull, advance head to drop the oldest item
            if state.tail.wrapping_sub(state.head) > capacity {
                state.head = state.head.wrapping_add(1);
            }
        }

        self.inner.write_notify.notify_one();
        Ok(data.len())
    }

    /// Adds a single element to the buffer.
    ///
    /// If the buffer is full, overwrites the oldest element.
    pub fn add(&self, item: T) -> Result<(), BufferError> {
        let mut state = self.inner.state.lock().unwrap();

        // Check for errors
        if let Some(ref err) = state.close_err {
            return Err(BufferError::ClosedWithError(Arc::clone(err)));
        }
        if state.close_write {
            return Err(BufferError::Closed);
        }

        let capacity = state.buf.len();
        let tail_idx = state.tail % capacity;
        state.buf[tail_idx] = Some(item);
        state.tail = state.tail.wrapping_add(1);

        // If we've exceeded capacity, advance head (overwrite oldest)
        if state.tail.wrapping_sub(state.head) > capacity {
            state.head = state.head.wrapping_add(1);
        }

        self.inner.write_notify.notify_one();
        Ok(())
    }

    /// Reads data from the buffer.
    ///
    /// Blocks when the buffer is empty until data becomes available.
    /// Returns the number of elements actually read.
    pub fn read(&self, buf: &mut [T]) -> Result<usize, BufferError> {
        let mut state = self.inner.state.lock().unwrap();

        // Check for errors
        if let Some(ref err) = state.close_err {
            return Err(BufferError::ClosedWithError(Arc::clone(err)));
        }

        // Wait for data
        while state.head == state.tail {
            if state.close_write {
                return Ok(0);
            }
            state = self.inner.write_notify.wait(state).unwrap();
            if let Some(ref err) = state.close_err {
                return Err(BufferError::ClosedWithError(Arc::clone(err)));
            }
        }

        let capacity = state.buf.len();
        let available = state.tail.wrapping_sub(state.head).min(capacity);
        let n = std::cmp::min(buf.len(), available);

        for i in 0..n {
            let head_idx = state.head % capacity;
            buf[i] = state.buf[head_idx].take().unwrap();
            state.head = state.head.wrapping_add(1);
        }

        Ok(n)
    }

    /// Returns the next element from the buffer (iterator pattern).
    ///
    /// Blocks when the buffer is empty until data becomes available.
    /// Returns `Err(Done)` when the buffer is closed and empty.
    pub fn next(&self) -> Result<T, Done> {
        let mut state = self.inner.state.lock().unwrap();

        // Check for errors
        if state.close_err.is_some() {
            return Err(Done);
        }

        // Wait for data
        while state.head == state.tail {
            if state.close_write {
                return Err(Done);
            }
            state = self.inner.write_notify.wait(state).unwrap();
            if state.close_err.is_some() {
                return Err(Done);
            }
        }

        // Read from head
        let capacity = state.buf.len();
        let head_idx = state.head % capacity;
        let item = state.buf[head_idx].take().unwrap();
        state.head = state.head.wrapping_add(1);

        Ok(item)
    }

    /// Discards the next n elements from the buffer.
    ///
    /// If n is greater than the number of available elements,
    /// all elements are discarded.
    pub fn discard(&self, n: usize) -> Result<(), BufferError> {
        let mut state = self.inner.state.lock().unwrap();
        if let Some(ref err) = state.close_err {
            return Err(BufferError::ClosedWithError(Arc::clone(err)));
        }

        let capacity = state.buf.len();
        let available = state.tail.wrapping_sub(state.head).min(capacity);
        let discard_count = std::cmp::min(n, available);

        for _ in 0..discard_count {
            let head_idx = state.head % capacity;
            state.buf[head_idx] = None;
            state.head = state.head.wrapping_add(1);
        }

        Ok(())
    }

    /// Returns a copy of all elements in the buffer.
    pub fn to_vec(&self) -> Vec<T> {
        let state = self.inner.state.lock().unwrap();
        let capacity = state.buf.len();
        let count = state.tail.wrapping_sub(state.head).min(capacity);
        let mut result = Vec::with_capacity(count);

        for i in 0..count {
            let idx = state.head.wrapping_add(i) % capacity;
            if let Some(ref item) = state.buf[idx] {
                result.push(item.clone());
            }
        }

        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::thread;
    use std::time::Duration;

    #[test]
    fn test_basic_write_read() {
        let buf = RingBuffer::<i32>::new(4);
        buf.write(&[1, 2, 3]).unwrap();

        let mut data = vec![0; 3];
        let n = buf.read(&mut data).unwrap();
        assert_eq!(n, 3);
        assert_eq!(data, vec![1, 2, 3]);
    }

    #[test]
    fn test_add_next() {
        let buf = RingBuffer::<i32>::new(4);
        buf.add(1).unwrap();
        buf.add(2).unwrap();
        buf.add(3).unwrap();

        assert_eq!(buf.next().unwrap(), 1);
        assert_eq!(buf.next().unwrap(), 2);
        assert_eq!(buf.next().unwrap(), 3);
    }

    #[test]
    fn test_overwrite_oldest() {
        let buf = RingBuffer::<i32>::new(3);

        // Fill buffer
        buf.write(&[1, 2, 3]).unwrap();
        assert_eq!(buf.len(), 3);

        // Write more - should overwrite oldest
        buf.write(&[4, 5]).unwrap();
        assert_eq!(buf.len(), 3);

        // Should get 3, 4, 5 (1 and 2 were overwritten)
        assert_eq!(buf.next().unwrap(), 3);
        assert_eq!(buf.next().unwrap(), 4);
        assert_eq!(buf.next().unwrap(), 5);
    }

    #[test]
    fn test_overwrite_more_than_capacity() {
        let buf = RingBuffer::<i32>::new(3);

        // Write more than capacity at once
        buf.write(&[1, 2, 3, 4, 5, 6, 7]).unwrap();

        // Should only have last 3 elements
        assert_eq!(buf.len(), 3);
        assert_eq!(buf.to_vec(), vec![5, 6, 7]);
    }

    #[test]
    fn test_close_write() {
        let buf = RingBuffer::<i32>::new(4);
        buf.add(1).unwrap();
        buf.close_write().unwrap();

        // Can still read existing data
        assert_eq!(buf.next().unwrap(), 1);

        // Now returns Done
        assert_eq!(buf.next(), Err(Done));

        // Cannot write after close
        assert!(buf.add(2).is_err());
    }

    #[test]
    fn test_capacity_and_len() {
        let buf = RingBuffer::<i32>::new(4);
        assert_eq!(buf.capacity(), 4);
        assert_eq!(buf.len(), 0);
        assert!(buf.is_empty());

        buf.write(&[1, 2, 3]).unwrap();
        assert_eq!(buf.len(), 3);
        assert!(!buf.is_empty());
    }

    #[test]
    fn test_concurrent_producer_consumer() {
        let buf = RingBuffer::<i32>::new(16);
        let producer_buf = buf.clone();

        let producer = thread::spawn(move || {
            for i in 0..100 {
                producer_buf.add(i).unwrap();
                thread::sleep(Duration::from_micros(10));
            }
            producer_buf.close_write().unwrap();
        });

        let mut collected = Vec::new();
        loop {
            match buf.next() {
                Ok(item) => collected.push(item),
                Err(Done) => break,
            }
        }

        producer.join().unwrap();
        // Due to overwriting, we might not get all 100
        assert!(!collected.is_empty());
    }

    #[test]
    fn test_reset() {
        let buf = RingBuffer::<i32>::new(4);
        buf.write(&[1, 2, 3]).unwrap();
        assert_eq!(buf.len(), 3);

        buf.reset();
        assert_eq!(buf.len(), 0);
        assert!(buf.is_empty());
    }

    #[test]
    fn test_discard() {
        let buf = RingBuffer::<i32>::new(8);
        buf.write(&[1, 2, 3, 4, 5]).unwrap();

        buf.discard(2).unwrap();
        assert_eq!(buf.len(), 3);
        assert_eq!(buf.next().unwrap(), 3);
    }

    #[test]
    fn test_to_vec() {
        let buf = RingBuffer::<i32>::new(4);
        buf.write(&[1, 2, 3]).unwrap();

        let data = buf.to_vec();
        assert_eq!(data, vec![1, 2, 3]);

        // Original buffer unchanged
        assert_eq!(buf.len(), 3);
    }

    #[test]
    fn test_sliding_window() {
        let buf = RingBuffer::<i32>::new(5);

        // Simulate streaming data
        for i in 0..20 {
            buf.add(i).unwrap();
        }

        // Should contain last 5 values
        let data = buf.to_vec();
        assert_eq!(data, vec![15, 16, 17, 18, 19]);
    }

    #[test]
    fn test_blocking_read() {
        let buf = RingBuffer::<i32>::new(4);
        let reader_buf = buf.clone();

        let reader = thread::spawn(move || {
            // This should block until data is available
            reader_buf.next().unwrap()
        });

        // Give the reader time to block
        thread::sleep(Duration::from_millis(10));

        // Write data to unblock the reader
        buf.add(42).unwrap();

        let result = reader.join().unwrap();
        assert_eq!(result, 42);
    }

    #[test]
    fn test_close_with_error() {
        let buf = RingBuffer::<i32>::new(4);
        buf.add(1).unwrap();

        let err = std::io::Error::new(std::io::ErrorKind::Other, "test error");
        buf.close_with_error(err).unwrap();

        // All operations return error
        assert!(buf.add(2).is_err());
        assert!(buf.read(&mut [0]).is_err());
    }

    #[test]
    fn test_error_method() {
        let buf = RingBuffer::<i32>::new(4);

        // No error initially
        assert!(buf.error().is_none());

        // After close_with_error, error() returns the error
        let err = std::io::Error::new(std::io::ErrorKind::Other, "test error");
        buf.close_with_error(err).unwrap();
        assert!(buf.error().is_some());
    }
}
